### ISSUE TITLE
Incorrect battery health in MacOS

### ISSUE THREAD

**Issue by WenqingZong** [ClementTsang/bottom#1164]
### Checklist

- [X] I've looked through [troubleshooting](https://clementtsang.github.io/bottom/nightly/troubleshooting), [known problems](https://clementtsang.github.io/bottom/nightly/support/official/#known-problems), and  [existing open issues](https://github.com/ClementTsang/bottom/issues?q=is%3Aopen+is%3Aissue) for similar issues.

### What operating system and version are you using?

MacOs 13.4

### What architecture are you using?

M1 chip

### What terminal(s) are you running bottom on?

Apple official Terminal app

### What version of bottom are you running?

0.9.1

### How did you install bottom?

Installed from brew

### Describe the issue

`bottom` failed to display the correct battery health information in MacOS.

![image](https://github.com/ClementTsang/bottom/assets/43934749/265e28d7-dcce-476f-9709-a216f6f3231d)

### What is the expected behaviour?

It's expected to be 100%, as shown in `System Settings`.

![image](https://github.com/ClementTsang/bottom/assets/43934749/a1eee4c8-33b9-48a5-b635-a15927445c67)

### What is the actual behaviour?

Battery health is shown as `2.28%`

### How can we reproduce this?

run `btm --battery` in Macbook

### Additional information

No

**Comment by ClementTsang** [comment]
Thanks, can reproduce. It's probably an issue with an upstream dependency, I'll investigate.

**Comment by ClementTsang** [comment]
Hm. Just some rambling as I debug this issue.

So I'm wondering if something's strange with what macOS reports - if I run `ioreg -l -w0 | grep Capacity` to get battery stats:

```
"MaxCapacity" = 100
"DesignCapacity" = 6075
"AppleRawMaxCapacity" = 5643
```

Which gives a calculated "battery health" of ~1.65 percent, if you divide the current max capacity by the designed capacity. This is how the library bottom uses calculates it, which you can see if you run `btm --battery`

<img width="342" alt="image" src="https://github.com/ClementTsang/bottom/assets/34804052/7d13a6ac-3afb-4d4b-bfd7-68ba45db4f1b">

Matches perfectly, 1.65%.

---

But as you noted, this is definitely wrong, if I check in settings, I have about 97% battery health. But I'm also not sure what I'm supposed to do then if what macOS is spitting out seems... wrong then? I could try using the raw max value which is closer, but that's still wrong (5643 / 6075 gives 93%, which still doesn't match what I see in Settings).

**Comment by ClementTsang** [comment]
And right as I post the comment, I think I figured out what I need... but I might need some extra information.

I think the columns I need are `AppleRawMaxCapacity` and `NominalChargeCapacity` - for me, this is:

```
"NominalChargeCapacity" = 5793
"AppleRawMaxCapacity" = 5643
```

5643 / 5793 = 97.4% - which is *very* close to what I see in options! So this might be it - if it is, I'll submit and/or fork a patch.

Could you help me confirm if this is the case by running `ioreg -l -w0 | grep Capacity` and giving me back:
- `NominalChargeCapacity`
- `AppleRawMaxCapacity`
- `MaxCapacity`
- `DesignCapacity`

Thanks!

**Comment by WenqingZong** [comment]
Hi, glad to receive your timely reply, here is what I get:

```text
NominalChargeCapacity 4507
AppleRawMaxCapacity 4377
MaxCapacity 100
DesignCapacity 4382
```

BTW, there is a `Health Information` section if you run `system_profiler detailLevel=mini SPPowerDataType`, but as it also reports `Maximum Capacity` so I'm not sure if it's what we need here.

**Comment by WenqingZong** [comment]
If it's possible, can I get the honour to fix this issue? I'm a recent graduate and seeking opportunities to build my reputation and contribution to FOSS.

**Comment by ClementTsang** [comment]
Go for it! I think the best place to look to make a change is https://github.com/starship/rust-battery to fix it upstream for everyone.

**Comment by WenqingZong** [comment]
Somehow I feel the bug is caused by different physics units. `rust-battery` represents `energy` in `joule`, but MacOS returns `energy` in `mAh`.

**Comment by WenqingZong** [comment]
Well, I got stuck at fixing it, and might need your help. My modifications cannot satisfy both our cases.

Your case:
```text
"MaxCapacity" = 100
"DesignCapacity" = 6075
"AppleRawMaxCapacity" = 5643
"NominalChargeCapacity" = 5793

and ground truth battery health: 97%
```

My case:
```text
AppleRawCurrentCapacity = 3338
AppleRawMaxCapacity = 4380
DesignCapacity = 4382
NominalChargeCapacity = 4515

and ground truth battery percentage: 80%, ground truth battery health: 100%
```

In file `rust-battery/src/platform/darwin/iokit/power_source.rs`, if I change 
```rust
static MAX_CAPACITY_KEY: &str = "MaxCapacity";
static CURRENT_CAPACITY_KEY: &str = "CurrentCapacity";
```
to
```rust
static MAX_CAPACITY_KEY: &str = "AppleRawMaxCapacity";
static CURRENT_CAPACITY_KEY: &str = "AppleRawCurrentCapacity";
```
then it match my battery health, but does not match my battery percentage and your battery health.

If, in addition to the previous two modifications, change:
```rust
static DESIGN_CAPACITY_KEY: &str = "DesignCapacity";
```
to
```rust
static DESIGN_CAPACITY_KEY: &str = "DesignCapacity";
```
Then, it does not match my percentage and health, but can match yours. :-(

Currently I have no idea why it bahaves in such a strange way. The only thing I can think of if I turned on [Charge on hold](https://discussions.apple.com/thread/252088720), but I don't think it makes a difference.

I'd appreciate it if you could give me some guidelines/ideas. Cheers.

**Comment by ClementTsang** [comment]
Some quick looking around gives [this thread](https://forums.macrumors.com/threads/m1-mba-battery-health-numbers-vary-from-different-sources.2287409/) which might be helpful to take a peek at. Note that whatever bottom/starship-battery is doing on at least M1, it's definitely wrong, so it does need to change for sure.

Also I'm curious if anyone has an Intel macbook to see if this is just an Apple Silicon issue or an overall macOS issue.

**Comment by WenqingZong** [comment]
Merge Request created: https://github.com/starship/rust-battery/pull/33

This fix only affects M-series-chip Mac as we have no idea if it's a bug for Intel-chip Mac.

I'll accept the fact that Apple might use some other data source to calculate battery percentage to deliver a better "user experience".

**Comment by ClementTsang** [comment]
Btw if upstream doesn't respond after a while, I can also look into forking the crate and including your fix to help get it out.

**Comment by WenqingZong** [comment]
> Btw if upstream doesn't respond after a while, I can also look into forking the crate and including your fix to help get it out.

The upstream approves my PR, waiting for someone to merge it.

**Comment by WenqingZong** [comment]
PR merged, and this bug will be fixed in the next version, `0.8.1`, of `rust-battery`. \o/

Happy to be a FOSS contributor :-)

**Comment by ClementTsang** [comment]
Thanks for fixing this! I'll make the relevant dependency bumps to include your fix (or if you want, you can also feel free to submit a PR), and I'll probably do a stable release sometime this week that fixes a few bugs.
